architecture-decision-records/0029-x-ray-for-request-monitoring.md

```diff
@@ -0,0 +1,61 @@
+* Cost: expensive to run due to being resource-intensive and commonly being paired with elastic search. AWS do a wrapped version called OpenSearch but this will be expensive to run as we need a fairly beefy server across multiple AZs
+
+### Use AWS X-Ray
+
+#### Advantages
+* Ease of setting up
+* Native Integration: AWS X-Ray seamlessly integrates with other AWS services, making it easy to monitor and trace requests across distributed systems within the AWS ecosystem.
+* Distributed Tracing: X-Ray provides end-to-end tracing capabilities, allowing you to visualize and understand the flow of requests through your application and identify bottlenecks or latency issues. 
+* AWS Service Instrumentation: X-Ray automatically instruments many AWS services, such as Lambda functions, API Gateway, and Elastic Load Balancer, providing insights into the performance of these services. 
+* AWS Ecosystem Support: X-Ray is well-suited for applications running on AWS, as it offers deep integration with other AWS monitoring and analytics services, such as CloudWatch and AWS Lambda.
+
+#### Disadvantages
+* Easy and convenient feature to use with AWS exclusive systems, the tool lacks when it comes to tracing outside the realms of Amazon’s products.
+* AWS-Centric: While X-Ray integrates well with AWS services, it may require additional effort to implement tracing for non-AWS resources or services. 
+* Limited Non-AWS Support: X-Ray's support for non-AWS services or environments is not as extensive compared to other monitoring solutions available in the market. 
+
+## Decision
+We have decided to use AWS-Xray for it's ease of setup and integration with other AWS services. In addition, other teams on AWS can use similar setup, and we can aggregate the X-Ray logs into a central account.
+Other bonuses include 
+* The first 100,000 traces recorded each month are free.
+* The first 1,000,000 traces retrieved or scanned each month are free.
+
+There are two main ways to send trace data to AWS.
+* Use the AWS X-Ray SDK. This needs quite a few code changes within the application to collect the data although there are plugins for Play and Akka that will make this easier. This doesn’t help much if we’re trying to run traces from Keycloak though. As part of this, we need to run a container with the X-Ray daemon.
+* Use the AWS Distro for open telemetry This is much easier to integrate in that you add a jar to the docker image when it’s being built and set some java options and it works out of the box. You do need to run another container with the amazon/aws-otel-collector docker image. If you want custom annotations on your traces, for example consignment ID so we can track individual consignments through the system, then you need a custom build of the collector docker image.  There is a frontend branch and a terraform branch which show roughly how this will work.
```

Comment:
The is an extra space before `There is a frontend branch and`.

Also, can you add a comma after `for example`?

Reply:
That last sentence isn't actually needed so I've removed it. I've added the comma after `for example`